### DISCUSSION TITLE
Enabling multiple plugins for different set of volumes as part of the same backup

### DISCUSSION THREAD

**Discussion by ashp7** [vmware-tanzu/velero#7465] [Community support Q&A]
I am new to velero and started exploring various options and plugins. have an use case where I have volumes across two backends and want to use two  different  backup plugins for these set of volumes as part of the same backup. Can someone help me with more information as to if It can be done and if so , how it can be done ?

appreciate your time and looking forward to more active participation in the community .

**Comment by sseago** [comment]
Are you talking about BackupItemAction plugins (which modify the content of the kubernetes metadata) or the native snapshot plugins (VolumeSnapshotter)?

If you mean BIA plugins, then as many plugins as match the resource will be invoked.
For the VolumeSnapshotter plugin, multiple plugins can be registered at once, but it would help to know your use case. For the supported Volume Snapshotters (aws, azure, and gcp), in most cases you won't have more than one of those storage types in use for the same cluster, since you'll have EBS volumes in AWS clusters, etc. -- but there may be cross-cloud storage use cases I'm unaware of.
If your volumes are CSI volumes, then there's only one CSI plugin, since any driver/provider-specific logic is not in the velero plugin but in the in-cluster CSI snapshotter that watches VolumeSnapshot resources.
If you mean that some volumes should use CSI snapshots and others native snapshots, I'm not 100% sure, but I think this might work as long as your cloud volumes (i.e. ebs volumes) don't use a CSI storageclass.

If you could provide a little more details around your use case, we might be able to give a clearer answer.

**Reply by ashp7** [reply]
Thank you so much for your reply.
Sure. let me try to provide bit more context. 

I want to perform a backup of resources spanning multiple PVCs across two providers - one eg.default provider like EBS and two would be a custom provider . Both of them support CSI snapshots and I expect volume snapshots would just work based on CSI. 
The difference in behavior I want to implement is to use default AWS plugins for performing backup to object store for EBS volumes and use custom plugin and implementation for performing backup to object store for the non-EBS volumes . 

I am not sure how and where to implement this. Does it require a new BIA plug-in ? If so how can I ensure that only certain volume use the custom plugin for the backup and the others use restic/kopia whichever is configured as default as  part of node agent

**Reply by ashp7** [reply]
@sseago Does my explanation help to provide more context ?

**Reply by sseago** [reply]
The object store plugin is used for sending Backup data to a specific BackupStorageLocation. This has nothing to do with the type of volume you are using -- it is used to store backup metadata, as well as volume data copied via kopia or restic. The volume type is really irrelevant in terms of which Object Store plugin you use. If your object store is an S3 object store, then the AWS plugin will be used. If it's an azure or gcp object store, then the azure or gcp plugin will be used. If you're using CSI snapshots without datamover, then none of the snapshot data is stored in the BackupStorageLocation, we just create the VolumeSnapshotContent. If you are using CSI snapshots *with* datamover, then after creating the VolumeSnapshot and VolumeSnapshotContent, Velero copies the data into the BackupStorageLocation using kopia. Since Velero is performing a filesystem copy from a PVC cloned from the Snapshot, it doesn't matter what volume type it is from the point of view of the Object Store/BackupStorageLocation.

You can create as many BackupStorageLocations as you need, but only one BSL will be used for a given backup. As long as your object store is S3-compliant (or is a GCP or Azure object store), you won't need to write a new ObjectStore plugin. In any case, BIA plugins aren't relevant -- those are used to manipulate Kubernetes metadata for individual items at backup time. Those plugins don't interact with storage back ends at all.
